editoast/editoast_schemas/src/rolling_stock.rs

```diff
@@ -87,3 +93,78 @@ pub struct RollingStock {
     #[serde(default)]
     pub metadata: Option<RollingStockMetadata>,
 }
+
+impl<'de> Deserialize<'de> for RollingStock {
```

Comment:
To make the implementation simpler, you might be able to use [this trick](https://stackoverflow.com/a/70293354) which allow to not have to maintain `Inner` with `RollingStock`.

```rust
#[derive(Deserialize)]
#[serde(remote = "self")]
struct RollingStock {
  /* ... */
}

impl<'de> Deserialize<'de> for RollingStock {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let rolling_stock = RollingStock::deserialize(deserializer)?;

        if rolling_stock
            .supported_signaling_systems
            .0
            .contains(&"ETCS_LEVEL2".to_string())
            && rolling_stock.etcs_brake_params.is_none()
        {
            return Err(serde::de::Error::custom(
                "invalid rolling-stock, supporting ETCS_LEVEL2 signaling system requires providing ETCS brake parameters.",
            ));
        }
        Ok(rolling_stock)
    }
}
```

Basically, `#[serde(remote = "self")]` doesn't really implement the `Deserialize` trait, but create a function of the object `RollingStock` ([this `remote` is actually here problems around the orphan rule](https://serde.rs/remote-derive.html)).